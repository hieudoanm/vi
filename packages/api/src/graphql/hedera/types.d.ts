export type Account = {
  email: string;
  accountId: string;
  publicKey: string;
  privateKey: string;
};

export type Balance = {
  symbol: string;
  name: string;
  balance: number;
};

export type Receipt = {
  accountId: string;
  tokenId: string;
  hbars: number;
  cents: number;
};

export type TopUpRequest = {
  amount: number;
  email: string;
};

export type TransferRequest = {
  amount: number;
  fromEmail: string;
  toEmail: string;
};
