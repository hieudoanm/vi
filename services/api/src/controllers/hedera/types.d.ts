export type AccountType = {
  email: string;
  accountId: string;
  publicKey: string;
  privateKey: string;
};

export type Balance = {
  symbol?: string;
  name?: string;
  balance?: number;
};

export type Receipt = {
  accountId: string;
  tokenId: string;
  hbars: number;
  cents: number;
};

export type TopUpRequestBody = {
  amount: number;
  email: string;
  tokenId?: string;
};

export type TransferRequestBody = {
  amount: number;
  fromEmail: string;
  toEmail: string;
  tokenId?: string;
};
