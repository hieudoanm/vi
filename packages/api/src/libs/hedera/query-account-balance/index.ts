import { AccountBalance, AccountBalanceQuery } from '@hashgraph/sdk';
import { VND_TOKEN_ID } from '../../../configs';
import getClient from '../client';

export const queryAccountBalance = async (
  accountId: string,
  tokenId: string = VND_TOKEN_ID
): Promise<AccountBalance> => {
  const client = getClient();

  const accountBalance: AccountBalance = await new AccountBalanceQuery()
    .setAccountId(accountId)
    .execute(client);

  console.log(
    `${accountId}: ${accountBalance.tokens?._map.get(
      tokenId
    )} units of token ID ${tokenId}`
  );

  return accountBalance;
};

export default queryAccountBalance;
