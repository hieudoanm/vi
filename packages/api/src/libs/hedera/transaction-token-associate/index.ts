import {
  PrivateKey,
  TokenAssociateTransaction,
  Transaction,
  TransactionReceipt,
} from '@hashgraph/sdk';
import getClient from '../client';
import execute from '../transaction-execute';

export const associateToken = async ({
  accountId,
  privateKey,
  tokenId,
}: {
  accountId: string;
  privateKey: string;
  tokenId: string;
}): Promise<TransactionReceipt> => {
  const client = getClient();

  const transaction: Transaction = await new TokenAssociateTransaction()
    .setAccountId(accountId)
    .setTokenIds([tokenId])
    .freezeWith(client)
    .sign(PrivateKey.fromStringED25519(privateKey));

  // EXECUTE
  const receipt: TransactionReceipt = await execute(transaction);

  return receipt;
};

export default associateToken;
