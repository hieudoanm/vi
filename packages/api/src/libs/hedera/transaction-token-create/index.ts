import {
  PrivateKey,
  TokenCreateTransaction,
  TokenSupplyType,
  TokenType,
  TransactionResponse,
} from '@hashgraph/sdk';
import getClient from '../client';

export const createToken = async (
  { initialSupply, decimals }: { initialSupply: number; decimals: number },
  { tokenCode, tokenName }: { tokenCode: string; tokenName: string },
  {
    treasuryAccountId,
    treasuryPrivateKey,
  }: { treasuryAccountId: string; treasuryPrivateKey: string }
) => {
  const client = getClient();

  const privateKey = PrivateKey.fromStringED25519(treasuryPrivateKey);

  const tokenCreateTransaction = await new TokenCreateTransaction()
    .setTokenName(tokenName)
    .setTokenSymbol(tokenCode)
    .setTokenType(TokenType.FungibleCommon)
    .setDecimals(decimals)
    .setInitialSupply(initialSupply)
    .setTreasuryAccountId(treasuryAccountId)
    .setAdminKey(privateKey)
    .setSupplyType(TokenSupplyType.Infinite)
    .setSupplyKey(privateKey)
    .freezeWith(client);

  // SIGN WITH TREASURY KEY
  const tokenCreateSign = await tokenCreateTransaction.sign(privateKey);

  // SUBMIT THE TRANSACTION
  const transactionResponse: TransactionResponse =
    await tokenCreateSign.execute(client);

  // GET THE TRANSACTION RECEIPT
  const transactionReceipt = await transactionResponse.getReceipt(client);

  // GET THE TOKEN ID
  const tokenId = transactionReceipt.tokenId;

  return tokenId;
};
