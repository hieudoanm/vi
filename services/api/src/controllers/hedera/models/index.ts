import mongoose, { Schema } from 'mongoose';

const AccountSchema = new Schema({
  email: { type: String, required: true, unique: true, index: true },
  accountId: { type: String, required: true, unique: true, index: true },
  publicKey: { type: String, required: true, unique: true, index: true },
  privateKey: { type: String, required: true, unique: true, index: true },
});

export const AccountModel = mongoose.model('Account', AccountSchema);
