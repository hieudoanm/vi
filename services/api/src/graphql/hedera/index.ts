export { resolvers } from './resolvers/hedera.resolver';
export { schemas } from './schemas/hedera.schema';
