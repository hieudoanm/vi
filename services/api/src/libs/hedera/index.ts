export { getClient } from './client';
export { queryAccountBalance } from './query-account-balance';
export { queryTokenInfo } from './query-token-info';
export { createAccount } from './transaction-create-account';
export { associateToken } from './transaction-token-associate';
export { transferToken } from './transaction-token-transfer';
