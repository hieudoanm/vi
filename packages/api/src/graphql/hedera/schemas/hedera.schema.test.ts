import { schemas } from './hedera.schema';

describe('hedera schemas', () => {
  it('should match snapshot', async () => {
    expect(schemas).toBeTruthy();
    expect(schemas).toMatchSnapshot();
  });
});
