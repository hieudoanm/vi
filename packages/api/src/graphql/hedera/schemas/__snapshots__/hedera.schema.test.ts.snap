// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hedera schemas should match snapshot 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "email",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "account",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Account",
            },
          },
        },
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "email",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "balance",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Balance",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeExtension",
      "name": Object {
        "kind": "Name",
        "value": "Query",
      },
    },
    Object {
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "email",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "account",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Account",
            },
          },
        },
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "amount",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "Float",
                  },
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "fromEmail",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "toEmail",
              },
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "transfer",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Transfer",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeExtension",
      "name": Object {
        "kind": "Name",
        "value": "Mutation",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "email",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "accountId",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "publicKey",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "privateKey",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Account",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "symbol",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "name",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "balance",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Int",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Balance",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "status",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Transfer",
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 469,
    "start": 0,
  },
}
`;
